cardinal/server/server.go

```diff
@@ -198,14 +203,33 @@ func (t *Handler) makeTxHandler(tx transaction.ITransaction) http.HandlerFunc {
 			writeError(writer, "unable to decode transaction", err)
 			return
 		}
-		t.w.AddTransaction(tx.ID(), txVal, sp)
 
-		res, err := json.Marshal("ok")
-		if err != nil {
-			writeError(writer, "unable to marshal response", err)
-			return
+		submitTx := func() uint64 {
+			tick := t.w.AddTransaction(tx.ID(), txVal, sp)
+
+			res, err := json.Marshal("ok")
+			if err != nil {
+				writeError(writer, "unable to marshal response", err)
+				return 0
+			}
+			writeResult(writer, res)
+			return uint64(tick)
+		}
+
+		// as long as we have an adapter, and the current world is not in recovery mode
+		if t.adapter != nil {
+			if t.w.IsRecovering() {
+				writeError(writer, "unable to submit transactions: game world is recovering state", nil)
+			} else {
+				tick := submitTx()
+				err = t.adapter.Submit(context.Background(), sp, uint64(tx.ID()), tick)
+				if err != nil {
+					writeError(writer, "error submitting transaction to blockchain", err)
+				}
+			}
+		} else {
+			submitTx()
 		}
```

Comment:
1. Can you quickly explan why the lack of an adapter in the handler means that the world should be in recovery mode?

2. The code seems to be checking if we have an adapter, and if we do, checks if the world `IsRecovering`, but I thought if we have an adapter, the world is not in recovery mode?

3. In which case should we call `submitTx()`, in recovery mode, or outside of recovery mode?

Reply:
the recovery mechanism reads transactions from the chain, which an adapter is needed to support. the recovery function won't run, and in turn won't set the `IsRecovering` value, unless an adapter is present in the world

if a transaction is submitted during recovery, we should not process it, as we need to get the game state back to where it was before, before we continue ticking

submit should only happen if we're not recovering

Reply:
ah, are you referring to this `// as long as we have an adapter, and the current world is not in recovery mode`? i can see how that is confusing. i can update it

Reply:
yeah the comment seemed contradictory